packages/python-packages/doc-warden/warden/enforce_changelog_content.py

```diff
@@ -0,0 +1,133 @@
+
+    for changelog_tuple in changelog_results:
+        if changelog_tuple[0] in known_issue_paths:
+            continue
+        if changelog_tuple[1]['curr_pkg_version'] != changelog_tuple[1]['latest_version_entry']:
+            missing_changelog.append(changelog_tuple)
+        elif len(changelog_tuple[1]['latest_release_notes']) == 0:
+            empty_release_notes.append(changelog_tuple)
+
+    return missing_changelog, empty_release_notes
+
+# parse rst to html, check for presence of appropriate version
+def verify_rst_changelog(changelog, config, pkg_version):
+    with open(changelog, 'r', encoding="utf-8") as f:
+        changelog_content = f.read()
+    html_changelog_content = rst_to_html(changelog_content)
+    html_soup = bs4.BeautifulSoup(html_changelog_content, "html.parser")
+
+    changelog_check_result = verify_latest_rst_section(html_soup, pkg_version)
+
+    return changelog, changelog_check_result
+
+# parse md to html, check for presence of the latest version
+def verify_md_changelog(changelog, config, pkg_version):
+    if config.verbose_output:
+        print('Examining content in {}'.format(changelog))
+
+    with open(changelog, 'r', encoding="utf-8") as f:
+        changelog_content = f.read()
+    html_changelog_content = markdown2.markdown(changelog_content)
+    html_soup = bs4.BeautifulSoup(html_changelog_content, "html.parser")
+
+    changelog_check_result = verify_latest_md_section(html_soup, pkg_version)
+
+    return changelog, changelog_check_result
+
+# get details from latest changelog entry
+def verify_latest_rst_section(html_soup, pkg_version):
+    changelog_check_result = {
+        'curr_pkg_version' : pkg_version,
+        'latest_version_entry' : '',
+        'latest_release_notes' : list()
+    }
+
+    latest_entry = html_soup.find('div', id='id1')
+    if latest_entry == None: latest_entry = html_soup.find('div', id='release-history')
+
+    if latest_entry != None:
+        for entry in latest_entry:
+            if entry.name == 'h1' or entry.name == 'h2':
+                changelog_check_result['latest_version_entry'] = entry.text.split(' ')[0]
+            if changelog_check_result['latest_version_entry'] == pkg_version:
+                if entry.name == 'p':
+                    changelog_check_result['latest_release_notes'].append(entry.text)
+                if entry.name == 'ul':
+                    for child in entry.children:
+                        if child.string != '\n': changelog_check_result['latest_release_notes'].append(child.string)
+
+    return changelog_check_result
+
+# get details from latest changelog entry
+def verify_latest_md_section(html_soup, pkg_version):
+    changelog_check_result = {
+        'curr_pkg_version' : pkg_version,
+        'latest_version_entry' : '',
+        'latest_release_notes' : list()
+    }
+
+    latest_version_tag = html_soup.find('h1', text=pkg_version)
```

Comment:
Why do you look for both h1 and h2 above, but only h1 here? 

IMO as long as it's an `<h>` tag we should examine it.

Reply:
the .rst files are formatted differently than the .md. For the markdown h2 (previously h1, but guidance changed) tags are used to specify the version. Then can be other levels of headers in between, but am just checking each h2 header and considering everything in between as release notes. Also I wonder how the changelog guidance affects the extraction of release tags, although I will test that out locally.